src/estimagic/parameters/nonlinear_constraints.py

```diff
@@ -0,0 +1,468 @@
+from functools import partial
+
+import numpy as np
+import pandas as pd
+from estimagic.differentiation.derivatives import first_derivative
+from estimagic.exceptions import InvalidFunctionError
+from estimagic.parameters.block_trees import block_tree_to_matrix
+from estimagic.parameters.tree_registry import get_registry
+from pybaum import tree_just_flatten
+from pybaum import tree_unflatten
+
+
+CONVERGENCE_ABSOLUTE_CONSTRAINT_TOLERANCE = 1e-5
+"""float: Allowed tolerance of the equality and inequality constraints for values to be
+considered 'feasible'.
+
+"""
+
+# ======================================================================================
+# Process nonlinear constraints
+# ======================================================================================
+
+
+def process_nonlinear_constraints(
+    nonlinear_constraints,
+    params,
+    converter,
+    numdiff_options,
+    skip_checks,
+):
+    """Process and prepare nonlinear constraints for internal use.
+
+    Args:
+        nonlinear_constraints (list[dict]): List of dictionaries, each representing a
+            nonlinear constraint.
+        params (pandas): A pytree containing the parameters with respect to which the
+            criterion is optimized. Examples are a numpy array, a pandas Series,
+            a DataFrame with "value" column, a float and any kind of (nested) dictionary
+            or list containing these elements. See :ref:`params` for examples.
+        converter (Converter): NamedTuple with methods to convert between internal and
+            external parameters, derivatives and function outputs.
+        numdiff_options (dict): Keyword arguments for the calculation of numerical
+            derivatives. See :ref:`first_derivative` for details. Note that the default
+            method is changed to "forward" for speed reasons. Contains lower and upper
+            bounds of parameters.
+        skip_checks (bool): Whether checks on the inputs are skipped. This makes the
+            optimization faster, especially for very fast constraint functions. Default
+            False.
+
+    Returns:
+        list[dict]: List of processed constraints.
+
+    """
+    constraint_evals = [
+        _check_validity_and_return_evaluation(c, params, skip_checks)
+        for c in nonlinear_constraints
+    ]
+
+    _processor = partial(
```

Comment:
I would not use partial just because a few arguments are constant if you use the function just once. Maybe use a regular loop instead of list comprehension then to keep it readable.